File: 0_50/20/20.py
class Solution:
    def isValid(self, s: str) -> bool:
        pairs = {
            '(': ')',
            '{': '}',
            '[': ']'
        }
        stack = []
        for c in s:
            if c in pairs:
                stack.append(c)
            elif not stack or c != pairs[stack.pop()]:
                return False
        return len(stack) == 0

File: 0000_50/20/20.dart
class Solution {
  bool isValid(String s) {
    Map<String, String> pairs = {'(': ')', '{': '}', '[': ']'};
    List<String> stack = [];
    for (int i = 0; i < s.length; i++) {
      if (pairs.containsKey(s[i]))
        stack.add(s[i]);
      else {
        if (stack.isEmpty || s[i] != pairs[stack.last]) return false;
        stack.removeLast();
      }
    }
    return stack.isEmpty;
  }
}

void main() {
  Solution S = Solution();
  String test = "(";
  print(S.isValid(test));
}
